### ISSUE TITLE
types_escaping_snapshot cleanup finalization

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#84570]
https://github.com/rust-lang/rust/pull/83185 deleted types_escaping_snapshot in compiler/rustc_infer/src/infer/type_variable.rs, but there are several comments remaining referencing that function, some of which seem to indicate no longer necessary operations. This code, to my knowledge, is pretty finicky, so we will want to be careful, but there may be some performance (or at least cleanup) wins left there.

cc @jyn514

**Comment by jyn514** [comment]
I'm not sure what's going on with that code. The comment on `impl sv::SnapshotVecDelegate for Delegate {` says 

> // In fact, we could *almost* just remove the
        // `SnapshotVec` entirely, except that we would have to
        // reproduce *some* of its logic, since we want to know which
        // type variables have been instantiated since the snapshot
        // was started, so we can implement `types_escaping_snapshot`.

but when I try to remove the impl I get lots of errors:
```
error[E0277]: the trait bound `type_variable::Delegate: SnapshotVecDelegate` is not satisfied
   --> compiler/rustc_infer/src/infer/type_variable.rs:319:10
    |
319 |     ) -> sv::SnapshotVec<Delegate, &mut Vec<TypeVariableData>, &mut InferCtxtUndoLogs<'tcx>> {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `SnapshotVecDelegate` is not implemented for `type_variable::Delegate`
    | 
   ::: /home/joshua/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/ena-0.14.0/src/snapshot_vec.rs:111:8
    |
111 |     D: SnapshotVecDelegate,
    |        ------------------- required by this bound in `SnapshotVec`
```
including from functions that are still used, like `values()`.

It was added in 57a593fcbb6af3db567c27b70b3a03c5a244705f - @nikomatsakis do any obvious cleanups come to mind now that types_escaping_snapshot no longer existsa?

**Comment by jyn514** [comment]
I did remove https://github.com/jyn514/rust/blob/6bed7e130942923e3e18d21e84121309c673e95e/compiler/rustc_infer/src/infer/type_variable.rs#L227 without any test failures.

**Comment by nikomatsakis** [comment]
Hmm

**Comment by nikomatsakis** [comment]
I don't think you can remove the `undo_log`, that's needed for probes and other things, but I might be forgetting something. There might be some simplifications possible.
